### PR TITLE
[libc] Provide isnan, isnanf and isnanl functions

### PR DESCRIPTION
**Pull request by petrhosek**
While C99 defines type generic isnan macro, BSD provided isnan, isnanf and isnanl in prior C standards and existing code still relies on these.

### CODE REVIEW

File: libc/include/math.h.def [hunk 1]
```diff
@@ -17,4 +17,6 @@
 
 %%public_api()
 
+#define isnan(x) __builtin_isnan(x)
```
Comment (petrhosek) [inline libc/include/math.h.def:20]: Should we put this macro into its own header? Do you have any suggestion for the name?

Comment (michaelrj-google) [reply libc/include/math.h.def:20]: If possible I'd like to leave this in `math-macros` but if it needs to be moved out then I'd say it should be defined in `isnan-macro.h` inside `include/llvm-libc-macros/`

Comment (jhuber6) [reply libc/include/math.h.def:20]: Do we expect every consumer of this header to have this builtin? We could possible do `if defined(__has_builtin) && __has_builtin(__builtin_isnan))` or omething.

Comment (lntue) [reply libc/include/math.h.def:20]: Does it mean that we still define `isnan` macro after defining all the `isnan` function?

If so then I think it makes sense to do the `TODO` in `math-macros.h`, that is to move `isinf`, `isfinite`, `isnan` macro definitions to `math-function-macros.h`, including it after `%%public_api` in here, and commenting about the reason we have to include it after all the function definitions.

**Review by nickdesaulniers — CHANGES_REQUESTED** [review]
`isnanf` and `isnanl` LGTM.

But I think we may need to be more careful here with `isnan`.

`isnan` is generally a macro or a function?  That needs to be explored more carefully I think.

cc @enh-google @frobtech for folks that might understand this problem better.

**Comment by enh-google** [comment]
> `isnan` is generally a macro or a function? That needs to be explored more carefully I think.

a macro, because realistically you want this to just be an inlined instruction or two via `__builtin_isnan()`. the c standard even explicitly says "7.12.3.5 The isnan macro" :-)

(callers that want the function [on systems where it exists] need to use the `(isnan)` trick. bionic and glibc do both provide a real function in addition to the macro, but iOS/macOS and musl don't.)

**Comment by petrhosek** [comment]
C standard introduced `isnan` as a macro in C99, but it has been provided as a function long before that, originally by BSD but it was also adopted by glibc etc. It's fine to define both as long as the macro is defined after the function which is what glibc, bionic and newlib do. I'm inclined to do the same for backwards compatibility.

**Comment by nickdesaulniers** [comment]
Got it, thanks for the context @enh-google . In that case, @petrhosek please add tests, but for isnan it sounds like we should test the function AND the macro?

File: libc/include/math.h.def (hunk 1 above)
Comment (nickdesaulniers) [reply libc/include/math.h.def:20]: > Do we expect every consumer of this header to have this builtin?

Yes.  All versions of clang and gcc that we claim to support have this builtin. Therefor, we should not guard it with __has_builtin (same for other builtins).

Comment (nickdesaulniers) [reply libc/include/math.h.def:20]: > that is to move isinf, isfinite, isnan

So isnan has isnan, isnanf, and isnanl.

isinf has isinf, isinff, isinfl,

finite has finite, finitef, finitel.  Note that finite is not isfinite{f|l}, so it lacks the symmetry that isinf and isnan have. Stated another way, isfinite (the type-generic macro) does not conflict with the finite, finitef, and finitel functions (like isnan and isinf do).

File: libc/config/baremetal/riscv/entrypoints.txt [hunk 2]
```diff
@@ -260,6 +260,9 @@ set(TARGET_LIBM_ENTRYPOINTS
     libc.src.math.ilogb
     libc.src.math.ilogbf
     libc.src.math.ilogbl
+    libc.src.math.isnan
+    libc.src.math.isnanf
+    libc.src.math.isnanl
```
Comment (nickdesaulniers) [inline libc/config/baremetal/riscv/entrypoints.txt:265]: We don't test baremetal continuously; please add these entrypoints to at least libc/config/linux/x86_64/entrypoints.txt, if not the rest of the linux targets.

**Comment by enh-google** [comment]
> C standard introduced `isnan` as a macro in C99, but it has been provided as a function long before that, originally by BSD but it was also adopted by glibc etc. It's fine to define both as long as the macro is defined after the function which is what glibc, bionic and newlib do. I'm inclined to do the same for backwards compatibility.

strictly, i don't think any of them (except bionic, for `isnan()` and `isinf()` only, not the f or l variants[1]) _declare_ the functions at all. there are just secret definitions, for backwards compatibility with a time when their macro turned into a call to one of the three (well, six) `__is*()` functions.

_____
1. and that seems like a mistake that i should fix by deletion.

**Comment by enh-google** [comment]
and glibc _does_ have all the declarations --- it just hides them behind macros that declare functions :-)

File: libc/include/math.h.def (hunk 1 above)
Comment (michaelrj-google) [reply libc/include/math.h.def:20]: For longer term maintainability I'd like to go with a design like what @lntue proposed where the primary math macros are included before `%%public_api` and the macros that might also be functions are defined in a separate header after.

Comment (petrhosek) [reply libc/include/math.h.def]: Done.

File: libc/config/baremetal/riscv/entrypoints.txt (hunk 2 above)
Comment (petrhosek) [reply libc/config/baremetal/riscv/entrypoints.txt:273]: Done.

**Review by lntue — APPROVED** [review]

**Comment by lntue** [comment]
Please fix formatting.

**Pull request merged**